Miniredis session tests appear to leak file descriptors

Now that we've added redis cluster tests in addition to standard and sentinel, we appear to have too many miniredis's trying to get going.  Without upping my `ulimit` past the default OSX setting of 256, I fail about 168 tests in of the 243 specs:

```
              Unexpected error:
                  <*net.OpError | 0xc000c90f00>: {
                      Op: "read",
                      Net: "tcp",
                      Source: {IP: [127, 0, 0, 1], Port: 65173, Zone: ""},
                      Addr: {IP: [127, 0, 0, 1], Port: 64908, Zone: ""},
                      Err: {},
                  }
                  read tcp 127.0.0.1:65173->127.0.0.1:64908: i/o timeout
              occurred
```

I even tried to reuse miniredis instances and ran `FlushAll` between specs, but then I eventually still got `socket: too many open files occurred` attempting to talk to the miniredis server.

## Expected Behavior

The tests work locally with sensible default system settings on most developers systems so they don't get tripped up by tests failing on the redis sessions.

## Current Behavior

Tests fail unless I run `ulimit -n 1024` to up my system resources limit.

## Possible Solution

I tried to reuse the same miniredis between all specs and flushed the redis data in between tests, but that didn't seem to fix the issue easily.  Still had issues with resources on the socket connections eventually at the same point.  So I'm out of quick and easy ideas.

I also tried to limit the max file descriptors with a buffered channel.  I couldn't get that to work well either.  I don't know how we get around systems being slow to release open file descriptors after a `miniredis.Close()` since they are likely sitting in a `TIME_WAIT` state or whatever it is for a bit of time afterwards.

## Steps to Reproduce (for bugs)

1.) ulimit -n 256
2.) go test ./...

I think its all redis tests because of the way it spins up a server on localhost and has a redis client speak to it.  I ditched all the sentinel & clusters tests and just did the standard tests in a loop 10 times.  Ran into the limit resource around the same point:

```
Ran 729 of 729 Specs in 0.454 seconds
FAIL! -- 254 Passed | 475 Failed | 0 Pending | 0 Skipped
--- FAIL: TestSessionStore (0.47s)
```

Since I also ran out of file descriptors when I tried reusing the same miniredis across all tests with DB flushes in between, my hunch is our issue is with the redis client itself in the session store, since a new one is spun up and grabs a new ephemeral port on its side of the TCP communication for every test.

When I tried using a channel to limit the new tests before an old test had closed, when I attempted to look at `mr.CurrentConnectionCount()` it always had 1.  Which I'm assuming is the client from the just finished spec. So maybe relying on standard garbage collection for each of those per spec is too delayed?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@
 
 ## Changes since v6.0.0
 
+- [#654](https://github.com/oauth2-proxy/oauth2-proxy/pull/654) Close client connections after each redis test (@JoelSpeed)
 - [#542](https://github.com/oauth2-proxy/oauth2-proxy/pull/542) Move SessionStore tests to independent package (@JoelSpeed)
 - [#577](https://github.com/oauth2-proxy/oauth2-proxy/pull/577) Move Cipher and Session Store initialisation out of Validation (@JoelSpeed)
 - [#635](https://github.com/oauth2-proxy/oauth2-proxy/pull/635) Support specifying alternative provider TLS trust source(s) (@k-wall)

--- a/pkg/sessions/redis/session_store_test.go
+++ b/pkg/sessions/redis/session_store_test.go
@@ -29,7 +29,14 @@ func TestSessionStore(t *testing.T) {
 }
 
 var _ = Describe("Redis SessionStore Tests", func() {
+	// helper interface to allow us to close client connections
+	// All non-nil redis clients should implement this
+	type closer interface {
+		Close() error
+	}
+
 	var mr *miniredis.Miniredis
+	var ss sessionsapi.SessionStore
 
 	BeforeEach(func() {
 		var err error
@@ -41,12 +48,25 @@ var _ = Describe("Redis SessionStore Tests", func() {
 		mr.Close()
 	})
 
+	JustAfterEach(func() {
+		// Release any connections immediately after the test ends
+		if redisStore, ok := ss.(*SessionStore); ok {
+			if redisStore.Client != nil {
+				Expect(redisStore.Client.(closer).Close()).To(Succeed())
+			}
+		}
+	})
+
 	tests.RunSessionStoreTests(
 		func(opts *options.SessionOptions, cookieOpts *options.CookieOptions) (sessionsapi.SessionStore, error) {
 			// Set the connection URL
 			opts.Type = options.RedisSessionStoreType
 			opts.Redis.ConnectionURL = "redis://" + mr.Addr()
-			return NewRedisSessionStore(opts, cookieOpts)
+
+			// Capture the session store so that we can close the client
+			var err error
+			ss, err = NewRedisSessionStore(opts, cookieOpts)
+			return ss, err
 		},
 		func(d time.Duration) error {
 			mr.FastForward(d)
@@ -63,7 +83,7 @@ var _ = Describe("Redis SessionStore Tests", func() {
 		})
 
 		AfterEach(func() {
-			go ms.Close()
+			ms.Close()
 		})
 
 		tests.RunSessionStoreTests(
@@ -74,7 +94,11 @@ var _ = Describe("Redis SessionStore Tests", func() {
 				opts.Redis.SentinelConnectionURLs = []string{sentinelAddr}
 				opts.Redis.UseSentinel = true
 				opts.Redis.SentinelMasterName = ms.MasterInfo().Name
-				return NewRedisSessionStore(opts, cookieOpts)
+
+				// Capture the session store so that we can close the client
+				var err error
+				ss, err = NewRedisSessionStore(opts, cookieOpts)
+				return ss, err
 			},
 			func(d time.Duration) error {
 				mr.FastForward(d)
@@ -90,7 +114,11 @@ var _ = Describe("Redis SessionStore Tests", func() {
 				opts.Type = options.RedisSessionStoreType
 				opts.Redis.ClusterConnectionURLs = []string{clusterAddr}
 				opts.Redis.UseCluster = true
-				return NewRedisSessionStore(opts, cookieOpts)
+
+				// Capture the session store so that we can close the client
+				var err error
+				ss, err = NewRedisSessionStore(opts, cookieOpts)
+				return ss, err
 			},
 			func(d time.Duration) error {
 				mr.FastForward(d)

